File: dexfile/src/main/java/com/github/netomi/bat/dexfile/visitor/DebugSequenceVisitor.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.visitor;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.debug.*;

public interface DebugSequenceVisitor
{
    default void visitAnyDebugInstruction(DexFile dexFile, DebugInfo debugInfo, DebugInstruction instruction) {
        throw new RuntimeException("Need to implement in class '" + this.getClass().getName() + "'.");
    }

    default void visitAdvanceLine(DexFile dexFile, DebugInfo debugInfo, DebugAdvanceLine instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitAdvanceLineAndPC(DexFile dexFile, DebugInfo debugInfo, DebugAdvanceLineAndPC instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitAdvancePC(DexFile dexFile, DebugInfo debugInfo, DebugAdvancePC instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitEndLocal(DexFile dexFile, DebugInfo debugInfo, DebugEndLocal instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitEndSequence(DexFile dexFile, DebugInfo debugInfo, DebugEndSequence instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitRestartLocal(DexFile dexFile, DebugInfo debugInfo, DebugRestartLocal instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitSetEpilogueBegin(DexFile dexFile, DebugInfo debugInfo, DebugSetEpilogueBegin instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitSetFile(DexFile dexFile, DebugInfo debugInfo, DebugSetFile instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitSetPrologueEnd(DexFile dexFile, DebugInfo debugInfo, DebugSetPrologueEnd instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitStartLocal(DexFile dexFile, DebugInfo debugInfo, DebugStartLocal instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }

    default void visitStartLocalExtended(DexFile dexFile, DebugInfo debugInfo, DebugStartLocalExtended instruction) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction);
    }
}

File: dexfile/src/main/kotlin/com/github/netomi/bat/dexfile/debug/visitor/DebugSequenceVisitor.kt
/*
 *  Copyright (c) 2020-2022 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.debug.visitor

import com.github.netomi.bat.dexfile.DexFile
import com.github.netomi.bat.dexfile.debug.*

fun interface DebugSequenceVisitor {
    fun visitAnyDebugInstruction(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugInstruction)

    fun visitAdvanceLine(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugAdvanceLine) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitAdvanceLineAndPC(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugAdvanceLineAndPC) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitAdvancePC(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugAdvancePC) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitEndLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugEndLocal) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitEndSequence(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugEndSequence) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitRestartLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugRestartLocal) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitSetEpilogueBegin(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugSetEpilogueBegin) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitSetFile(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugSetFile) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitSetPrologueEnd(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugSetPrologueEnd) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitStartLocal(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugStartLocal) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }

    fun visitStartLocalExtended(dexFile: DexFile, debugInfo: DebugInfo, instruction: DebugStartLocalExtended) {
        visitAnyDebugInstruction(dexFile, debugInfo, instruction)
    }
}